apps/worker/src/app/workflow/usecases/send-message/send-message-push.usecase.ts

```diff
@@ -113,81 +120,103 @@ export class SendMessagePush extends SendMessageBase {
         Object.values(PushProviderIdEnum).includes(chan.providerId as PushProviderIdEnum)
       ) || [];
 
-    const messagePayload = Object.assign({}, command.payload);
-    delete messagePayload.attachments;
-
     if (!pushChannels.length) {
-      await this.sendErrors(command);
+      await this.sendNoActiveChannelError(command.job);
+      await this.sendNotificationError(command.job);
 
       return;
     }
 
-    for (const channel of pushChannels) {
-      if (!channel.credentials?.deviceTokens) {
-        await this.createExecutionDetails.execute(
-          CreateExecutionDetailsCommand.create({
-            ...CreateExecutionDetailsCommand.getDetailsFromJob(command.job),
-            detail: DetailEnum.PUSH_MISSING_DEVICE_TOKENS,
-            source: ExecutionDetailsSourceEnum.INTERNAL,
-            status: ExecutionDetailsStatusEnum.FAILED,
-            isTest: false,
-            isRetry: false,
-          })
-        );
-
-        continue;
-      }
+    const messagePayload = Object.assign({}, command.payload);
+    delete messagePayload.attachments;
 
-      const integration = await this.getIntegration({
-        id: channel._integrationId,
-        organizationId: command.organizationId,
-        environmentId: command.environmentId,
-        channelType: ChannelTypeEnum.PUSH,
-        providerId: channel.providerId,
-        userId: command.userId,
-        filterData: {
-          tenant: command.job.tenant,
-        },
-      });
+    let integrationsWithErrors = 0;
+    for (const channel of pushChannels) {
+      const { deviceTokens } = channel.credentials || {};
 
-      if (!integration) {
-        await this.createExecutionDetails.execute(
-          CreateExecutionDetailsCommand.create({
-            ...CreateExecutionDetailsCommand.getDetailsFromJob(command.job),
-            detail: DetailEnum.SUBSCRIBER_NO_ACTIVE_INTEGRATION,
-            source: ExecutionDetailsSourceEnum.INTERNAL,
-            status: ExecutionDetailsStatusEnum.FAILED,
-            isTest: false,
-            isRetry: false,
-          })
-        );
+      const [isChannelMissingDeviceTokens, integration] = await Promise.all([
```

Comment:
Now we can execute this paralelly as we do not break the loop anymore, nor neither want to break it.

Reply:
I considered it while thinking how we should manage all these cases but I went this direction to show everything that is failing in once.

Now that you raise the number of calls we need to make, thinking about it I can see this solution should executing less calls in an error case situation based on this generic scenario:
- All the amount of db calls we do to trigger a notification (let's call it N pre-calls).
- In SendMessagePush we fail because missing device tokens.
- User fixes that and sends again N pre-calls.
- In SendMessagePush the Redis SelectIntegration call returns an user error. (+1 call)
- User fixes that and sends again N pre-calls.
- Notification is being sent successfully with SendMessagePush X post-calls.

vs

- N pre-calls
- In SendMessagePush we fail because missing device tokens and select integration (+1 call)
- User fixes those and sends again N pre-calls.
- Notification is being sent successfully with SendMessagePush X post-calls.

For this kind of scenario makes:
3 * N pre-calls + 1 + X vs 2 * N pre-calls + 1 + X

Also the less trigger events generated the less load our systems will have and the less the customers will pay.

Maybe if we start to add more conditions on that phase of the flow we can incur in extra SelectIntegration calls while are only other preconditions failing. So I can't predict in total numbers how our users will not set properly the configuration, but I am guessing that statistically should be better to do an extra call that having to do a set of N pre-calls if the select integration brings that the subscriber got no integrations assigned.
I would be happy to brainstorm the flow with you to improve and make it more performant if you want.

Reply:
I see what you mean here and it is a possible use case that I was not thinking of.
The reason I raised this comment was because this 'error' is on the job-subscriber-channel level and not the whole notification (all the workflow flow). 
If there is a case where the user sees that he has X subscribers that failed to send push messages from N subscribers, IMO he should resent the notification only for X subscribers and potentially only for this step, otherwise these subscribers will get duplicated messages.

Anyway, those two cases should be pretty rare so I am good with any of them :)